File: java/src/LC_5_LongestPalindromicSubstring.java
// https://leetcode.com/problems/longest-palindromic-substring

public class LC_5_LongestPalindromicSubstring {

    public int expandAroundCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        return right - left - 1;
    }

    public void lenOfLongestPalindromicSubstring(String s) {
        int start = 0, end = 0;
        for (int i = 0; i < s.length(); i++) {

            int len1 = expandAroundCenter(s, i, i);
            int len2 = expandAroundCenter(s, i, i + 1);
            int len = Math.max(len1, len2);
            if (len > end - start) {
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }

        System.out.println(s.substring(start, end + 1));

    }

    public static void main(String[] args) {
        LC_5_LongestPalindromicSubstring solution = new LC_5_LongestPalindromicSubstring();
        solution.lenOfLongestPalindromicSubstring("babad");
        solution.lenOfLongestPalindromicSubstring("cbbd");
        solution.lenOfLongestPalindromicSubstring("madam");
    }
}

File: python/two_pointers/LC_5_LongestPalindromicSubstring.py
# https://leetcode.com/problems/longest-palindromic-substring
class Solution:

    def expand_around_center(self, s, left, right):
        while (left >= 0 and right < len(s) and s[left] == s[right]):
            left -= 1
            right += 1
        return right-left-1

    def longestPalindrome(self, s: str) -> str:
        start, end = 0., 0
        for i in range(len(s)):
            len1 = self.expand_around_center(s, i, i)
            len2 = self.expand_around_center(s, i, i+1)
            sp_len = max(len1, len2)
            if sp_len > end-start:
                start = i - (sp_len-1)//2
                end = i + sp_len//2
        print(s[start:end+1])
        return s[start:end+1]


sol = Solution()
assert sol.longestPalindrome("babad") == "aba"
